### PR TITLE
7873 Send a message with a link to the order by email when the order status is 'unpaid'

### PR DESCRIPTION
**Pull request by Cr1stal423**
GreenCityUBS PR
Issue Link:
https://github.com/orgs/ita-social-projects/projects/43/views/1?pane=issue&itemId=89867663&issue=ita-social-projects%7CGreenCity%7C7873

Changes:

    Added new notification logic
    Added a few methods to check if the order is unpaid and send an email
    Added asynchronous processing

Current behavior:

If the order is not paid for, the user immediately receives an email notification about it, with a link to the order

image

### CODE REVIEW

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 1]
```diff
@@ -613,6 +609,25 @@ public PaymentSystemResponse saveFullOrderToDB(OrderResponseDto dto, String uuid
         return processPayment(dto, order, sumToPayInCoins, currentUser);
     }
 
+    @Async
+    public void checkIfOrderIsNotPayedAndSendEmailAsync(Order order, Long sumToPayInCoins) {
+        try {
+            Thread.sleep(1000);
```
Comment (LazarenkoDmytro) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: Please replace the Thread.sleep() with the Spring TaskScheduler. The sleep() does not align with the asynchronous execution, blocking the thread and making it do nothing. TaskScheduler, on the other hand, schedules the task in one second, making the thread available for other tasks

Comment (Cr1stal423) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: I'll deal with it

**Review by Maryna-511750 — APPROVED** [review]

File: service-api/src/main/java/greencity/service/ubs/UBSClientService.java [hunk 2]
```diff
@@ -3,7 +3,7 @@
 import greencity.dto.CreateAddressRequestDto;
 import greencity.dto.LocationsDto;
 import greencity.dto.TariffInfoByLocationDto;
-import greencity.dto.order.OrderAddressExportDetailsDtoUpdate;
+import greencity.dto.order.*;
```
Comment (KizerovDmitriy) [inline service-api/src/main/java/greencity/service/ubs/UBSClientService.java]: we don't use wildcard imports

Comment (Cr1stal423) [reply service-api/src/main/java/greencity/service/ubs/UBSClientService.java]: Fixed

File: core/src/main/java/greencity/controller/OrderController.java [hunk 3]
```diff
@@ -60,6 +60,8 @@
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.context.request.RequestAttributes;
+import org.springframework.web.context.request.RequestContextHolder;
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/OrderController.java]: Unused imports

File: service-api/src/main/java/greencity/service/ubs/NotificationService.java [hunk 4]
```diff
@@ -233,4 +233,14 @@ PageableDto<NotificationShortDto> getAllNotificationsForUser(String userUuid,
      * @author Roman Kasarab
      */
     void deleteNotification(Long notificationId, String userUuid);
+
+    /**
+     * Notify user that order has unpaid status for a long time. This method is used
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/service/ubs/NotificationService.java]: what does `a long time` mean here?

File: service-api/src/main/java/greencity/service/ubs/NotificationService.java [hunk 5]
```diff
@@ -233,4 +233,14 @@ PageableDto<NotificationShortDto> getAllNotificationsForUser(String userUuid,
...
+     * by schedule task, which runs every day.
+     *
+     * @param order the order to send notification for
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/service/ubs/NotificationService.java]: please add `sumToPay` to the java doc

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java [hunk 6]
```diff
@@ -787,6 +787,16 @@ public void deleteNotification(Long notificationId, String userUuid) {
         userNotificationRepository.markAsDeletedUserNotificationByIdAndUserId(notificationId, userId);
     }
 
+    @Override
+    public void notifyUnpaidOrderPermanently(Order order, Long amountToPay) {
+        boolean isOrderPayed = order.getOrderPaymentStatus().equals(OrderPaymentStatus.PAID);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/notification/NotificationServiceImpl.java]: you can use `==` to compare enum values

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java [hunk 7]
```diff
@@ -787,6 +787,16 @@ public void deleteNotification(Long notificationId, String userUuid) {
...
+        if (!isOrderPayed) {
+            Double amount = amountToPay.doubleValue() / 100.0;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/notification/NotificationServiceImpl.java]: why do we need this? 
please move `100.0` to the constant

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java [hunk 8]
```diff
@@ -787,6 +787,16 @@ public void deleteNotification(Long notificationId, String userUuid) {
         userNotificationRepository.markAsDeletedUserNotificationByIdAndUserId(notificationId, userId);
     }
 
+    @Override
+    public void notifyUnpaidOrderPermanently(Order order, Long amountToPay) {
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/notification/NotificationServiceImpl.java:792]: based on the java doc this method should be run `by schedule task, which runs every day.`
But I see that it is called in `saveFullOrderToDB`

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 9]
```diff
@@ -267,6 +276,10 @@ public class UBSClientServiceImpl implements UBSClientService {
     private final CityRepository cityRepository;
     private final DistrictRepository districtRepository;
     private final MonoBankClient monoBankClient;
+    private final TaskScheduler taskScheduler;
+    private final NotificationServiceImpl notificationServiceImpl;
+    @PersistenceContext
+    private EntityManager entityManager;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: why do we need it?
it is never used

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 10]
```diff
@@ -267,6 +276,10 @@ public class UBSClientServiceImpl implements UBSClientService {
     private final CityRepository cityRepository;
     private final DistrictRepository districtRepository;
     private final MonoBankClient monoBankClient;
+    private final TaskScheduler taskScheduler;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: why do we need it?
it is never used

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 11]
```diff
@@ -164,6 +171,7 @@
 import java.util.Base64;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Date;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: unused  import

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 12]
```diff
@@ -132,20 +132,27 @@
...
 import greencity.service.locations.LocationApiService;
+import greencity.service.notification.NotificationServiceImpl;
 import greencity.service.phone.UAPhoneNumberUtil;
 import greencity.util.Bot;
 import greencity.util.EncryptionUtil;
 import greencity.util.OrderUtils;
+import jakarta.persistence.EntityManager;
 import jakarta.persistence.EntityNotFoundException;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.TableGenerator;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: unused import

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 13]
```diff
@@ -230,6 +238,7 @@
  */
 @Service
 @RequiredArgsConstructor
+@Slf4j
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: why do we need it?
I do not see any logging in this class

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java (hunk 8 above)
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/notification/NotificationServiceImpl.java:792]: based on the java doc this method should be run `by schedule task, which runs every day.`
But I see that it is called in `saveFullOrderToDB`

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java (hunk 10 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java]: I just forgot to delete everything related to fields and imports that are not used. I have already fixed everything

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java (hunk 8 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/notification/NotificationServiceImpl.java:792]: Changed java doc

File: service-api/src/main/java/greencity/service/ubs/NotificationService.java (hunk 4 above)
Comment (Cr1stal423) [reply service-api/src/main/java/greencity/service/ubs/NotificationService.java]: The Java doc was written by JetBrains AI😁, he did a bad job, and I changed the description myself

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java (hunk 6 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/notification/NotificationServiceImpl.java]: Changed

File: service/src/main/java/greencity/service/notification/NotificationServiceImpl.java (hunk 7 above)
Comment (Cr1stal423) [reply service/src/main/java/greencity/service/notification/NotificationServiceImpl.java]: A good point, I've already made it a constant

**Review by holotsvan — APPROVED** [review]

**Review by RomaMocherniuk — APPROVED** [review]

**Review by KizerovDmitriy — APPROVED** [review]

**Pull request merged**